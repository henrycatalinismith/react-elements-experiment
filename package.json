{
  "name": "@hendotcat/elements",
  "description": "React component library",
  "version": "0.0.0",
  "license": "MIT",
  "homepage": "https://github.com/hendotcat/elements",
  "author": {
    "name": "Henry Catalini Smith",
    "email": "smith@hen.cat",
    "url": "https://hen.cat/"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/hendotcat/elements.git"
  },
  "devDependencies": {
    "@types/jest": "^26.0.23",
    "@types/react": "^17.0.6",
    "@types/react-dom": "^17.0.5",
    "jest": "^26.6.3",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "rehype": "^11.0.0",
    "rehype-format": "^3.1.0",
    "ts-jest": "^26.5.6",
    "typescript": "^4.2.4"
  },
  "peerDependencies": {
    "react": "^17.0.2"
  },
  "scripts": {
    "build": "tsc"
  },
  "files": [
    "elements.d.ts",
    "elements.js"
  ],
  "main": "elements.js",
  "types": "elements.d.ts",
  "jest": {
    "transform": {
      "^.+\\.(ts)x?$": "ts-jest"
    }
  }
}
