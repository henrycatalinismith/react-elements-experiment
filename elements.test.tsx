import { select } from "hast-util-select"
import React from "react"
import ReactDOMServer from "react-dom/server"
import rehype from "rehype"
import rehypeDomParse from "rehype-dom-parse"
import rehypeFormat from "rehype-format"
import unified from "unified"
import { name, version } from "./package.json"
import {
  Body,
  Div,
  Document,
  H1,
  Head,
  Heading,
  Span,
} from "./elements"

function render(component: any): string {
  return rehype()
    .use(rehypeFormat)
    .processSync(ReactDOMServer.renderToString(component))
    .toString()
}

function parse(html: string): any {
  return unified()
    .use(rehypeDomParse, { fragment: false })
    .parse(html)
}

describe(`${name}@${version}`, () => {
  describe("<Document />", () => {
    it("renders a valid document given the bare minimum input", () => {
      expect(render(
        <Document lang="en-US" title="test" />
      )).toMatchSnapshot()
    })

    it("renders custom header metadata", () => {
      expect(render(
        <Document lang="en-US" title="test">
          <Head>
            <meta
              name="viewport"
              content="width=device-width, initial-scale=1"
            />
          </Head>
        </Document>
      )).toMatchSnapshot()
    })

    it("treats bare children as body contents", () => {
      expect(render(
        <Document lang="en-US" title="test">
          <h1>title</h1>
        </Document>
      )).toMatchSnapshot()
    })

    it("renders <Body> children as body contents", () => {
      expect(render(
        <Document lang="en-US" title="test">
          <Body>
            <h1>title</h1>
          </Body>
        </Document>
      )).toMatchSnapshot()
    })
  })

  describe("<H1 />", () => {
    it("renders a <h1>", () => {
      expect(render(
        <Document lang="en-US" title="test">
          <H1>title</H1>
        </Document>
      )).toMatchSnapshot()
    })
  })

  describe("<Heading />", () => {
    it("renders a <h1> by default", () => {
      expect(render(
        <Document lang="en-US" title="test">
          <Heading>title</Heading>
        </Document>
      )).toMatchSnapshot()
    })

    it("allows heading level changes of up to 1", () => {
      expect(render(
        <Document lang="en-US" title="test">
          <Heading level={1}>one</Heading>
          <Heading level={2}>two</Heading>
          <Heading level={3}>three</Heading>
          <Heading level={4}>four</Heading>
          <Heading level={5}>five</Heading>
          <Heading level={6}>six</Heading>
          <Heading level={5}>five</Heading>
          <Heading level={4}>four</Heading>
          <Heading level={3}>three</Heading>
          <Heading level={2}>two</Heading>
          <Heading level={1}>one</Heading>
        </Document>
      )).toMatchSnapshot()
    })

    it("rejects heading level changes of more than 1", () => {
      expect(() => render(
        <Document lang="en-US" title="test">
          <Heading level={1}>one</Heading>
          <Heading level={3}>three</Heading>
        </Document>
      )).toThrowErrorMatchingSnapshot()
    })
  })

  describe("<LanguageProvider />", () => {
    it("includes lang attribute if different from document", () => {
      const html = render(
        <Document lang="en-US" title="test">
          <Heading lang="sv-SE">titel</Heading>
        </Document>
      )
      const tree = parse(html)
      expect(select("html[lang='en-US']", tree)).toBeTruthy()
      expect(select("h1[lang='sv-SE']", tree)).toBeTruthy()
    })

    it("excludes lang attribute if same as document", () => {
      const html = render(
        <Document lang="en-US" title="test">
          <Heading lang="en-US">title</Heading>
        </Document>
      )
      const tree = parse(html)
      expect(select("html[lang='en-US']", tree)).toBeTruthy()
      expect(select("h1:not([lang='en-US'])", tree)).toBeTruthy()
    })

    it("includes lang attribute if different from parent", async () => {
      const html = render(
        <Document lang="en-US" title="test">
          <Heading lang="sv-SE">
            <Span lang="en-US">
              title
            </Span>
          </Heading>
        </Document>,
      )
      const tree = parse(html)
      expect(select("html[lang='en-US']", tree)).toBeTruthy()
      expect(select("h1[lang='sv-SE']", tree)).toBeTruthy()
      expect(select("span[lang='en-US']", tree)).toBeTruthy()
    })
  })

  describe("<LevelProvider />", () => {
    it("allows inline elements inside block elements", () => {
      expect(() => render(
        <Document lang="en-US" title="test">
          <Div><Span>test</Span></Div>
        </Document>
      )).not.toThrowError()
    })

    it("disallows block elements inside inline elements", () => {
      expect(() => render(
        <Document lang="en-US" title="test">
          <Span><Div>test</Div></Span>
        </Document>
      )).toThrowError()
    })
  })
})
