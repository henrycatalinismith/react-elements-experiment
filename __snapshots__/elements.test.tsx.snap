// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@hendotcat/elements <DescriptionList /> renders a list of links with descriptions 1`] = `
"
<html lang=\\"en-US\\">
  <head>
    <meta charset=\\"utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"description\\" content=\\"testing\\">
    <title>test</title>
  </head>
  <body>
    <dl>
      <dt><a href=\\"https://example.org/\\">link</a></dt>
      <dd><span>an example website</span></dd>
    </dl>
  </body>
</html>
"
`;

exports[`@hendotcat/elements <Document /> renders <Body> children as body contents 1`] = `
"
<html lang=\\"en-US\\">
  <head>
    <meta charset=\\"utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"description\\" content=\\"testing\\">
    <title>test</title>
  </head>
  <body>
    <h1>title</h1>
  </body>
</html>
"
`;

exports[`@hendotcat/elements <Document /> renders a valid document given the bare minimum input 1`] = `
"
<html lang=\\"en-US\\">
  <head>
    <meta charset=\\"utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"description\\" content=\\"testing\\">
    <title>test</title>
  </head>
  <body></body>
</html>
"
`;

exports[`@hendotcat/elements <Document /> renders custom header metadata 1`] = `
"
<html lang=\\"en-US\\">
  <head>
    <meta charset=\\"utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"description\\" content=\\"testing\\">
    <title>test</title>
    <meta property=\\"og:title\\" content=\\"test\\">
  </head>
  <body></body>
</html>
"
`;

exports[`@hendotcat/elements <Document /> treats bare children as body contents 1`] = `
"
<html lang=\\"en-US\\">
  <head>
    <meta charset=\\"utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"description\\" content=\\"testing\\">
    <title>test</title>
  </head>
  <body>
    <h1>title</h1>
  </body>
</html>
"
`;

exports[`@hendotcat/elements <H1 /> renders a <h1> 1`] = `
"
<html lang=\\"en-US\\">
  <head>
    <meta charset=\\"utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"description\\" content=\\"testing\\">
    <title>test</title>
  </head>
  <body>
    <h1>title</h1>
  </body>
</html>
"
`;

exports[`@hendotcat/elements <Heading /> allows heading level changes of up to 1 1`] = `
"
<html lang=\\"en-US\\">
  <head>
    <meta charset=\\"utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"description\\" content=\\"testing\\">
    <title>test</title>
  </head>
  <body>
    <h1>one</h1>
    <h2>two</h2>
    <h3>three</h3>
    <h4>four</h4>
    <h5>five</h5>
    <h6>six</h6>
    <h5>five</h5>
    <h4>four</h4>
    <h3>three</h3>
    <h2>two</h2>
    <h1>one</h1>
  </body>
</html>
"
`;

exports[`@hendotcat/elements <Heading /> rejects heading level changes of more than 1 1`] = `"Heading level 3 not allowed: previous level is 1"`;

exports[`@hendotcat/elements <Heading /> renders a <h1> by default 1`] = `
"
<html lang=\\"en-US\\">
  <head>
    <meta charset=\\"utf-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"description\\" content=\\"testing\\">
    <title>test</title>
  </head>
  <body>
    <h1>title</h1>
  </body>
</html>
"
`;
